File: tests-python/pages/login_page.py
from __future__ import annotations

import allure
from selenium.webdriver.common.by import By

from pages.base import BasePage


class LoginPage(BasePage):
    LOGIN_INPUT = (By.CSS_SELECTOR, "[data-testid='login-input']")
    PASSWORD_INPUT = (By.CSS_SELECTOR, "[data-testid='password-input']")
    SUBMIT = (By.CSS_SELECTOR, "[data-testid='submit-button']")
    FORM_TITLE = (By.CSS_SELECTOR, "[data-testid='login-form-title']")
    ERROR = (By.CSS_SELECTOR, "[data-testid='error-message']")
    REGISTER_LINK = (By.CSS_SELECTOR, "[data-testid='register-link']")

    @allure.step("Open login page")
    def open_page(self) -> LoginPage:
        self.open_path("/login")
        return self

    @allure.step("Click 'Register' link under the login form")
    def click_register_link(self):
        from pages.register_page import RegisterPage

        self.find(*self.REGISTER_LINK).click()
        return RegisterPage(self.driver, self.config)

    @allure.step("Fill and submit form")
    def fill_and_submit_form(self, username: str, password: str):
        self.type_username(username)
        self.type_password(password)
        return self.submit()

    @allure.step("Type username: {username}")
    def type_username(self, username: str) -> LoginPage:
        self.fill(*self.LOGIN_INPUT, username)
        return self

    @allure.step("Type password")
    def type_password(self, password: str) -> LoginPage:
        self.fill(*self.PASSWORD_INPUT, password)
        return self

    @allure.step("Submit login form")
    def submit(self):
        from pages.home_page import HomePage

        self.find(*self.SUBMIT).click()
        self.wait_url_is_home()
        return HomePage(self.driver, self.config)

    @allure.step("Submit login form expecting validation error")
    def submit_expecting_error(self) -> LoginPage:
        self.find(*self.SUBMIT).click()
        self.wait_visible(*self.ERROR)
        return self

    @allure.step("Verify login form is mounted")
    def should_show_login_form(self) -> LoginPage:
        self.wait_visible(*self.FORM_TITLE)
        self.wait_visible(*self.LOGIN_INPUT)
        self.wait_visible(*self.PASSWORD_INPUT)
        self.wait_visible(*self.SUBMIT)
        return self

    @allure.step("Verify form title message: {message}")
    def should_have_form_title(self, message: str) -> LoginPage:
        el = self.wait_visible(*self.FORM_TITLE)
        assert message in el.text
        return self

    @allure.step("Verify error message: {message}")
    def should_have_error_message(self, message: str) -> LoginPage:
        self.wait_text_contains(*self.ERROR, message)
        return self

File: java/gradle-junit5-selenium/src/test/java/pages/LoginPage.java
package pages;

import config.TestConfig;
import io.qameta.allure.Step;
import org.openqa.selenium.By;
import org.openqa.selenium.WebDriver;

public class LoginPage extends BasePage {

    private static final By LOGIN_INPUT = By.cssSelector("[data-testid='login-input']");
    private static final By PASSWORD_INPUT = By.cssSelector("[data-testid='password-input']");
    private static final By SUBMIT = By.cssSelector("[data-testid='submit-button']");
    private static final By FORM_TITLE = By.cssSelector("[data-testid='login-form-title']");
    private static final By ERROR = By.cssSelector("[data-testid='error-message']");
    private static final By REGISTER_LINK = By.cssSelector("[data-testid='register-link']");
    private static final By HEADER = By.cssSelector("[data-testid='header']");

    public LoginPage(WebDriver driver, TestConfig config) {
        super(driver, config);
    }

    @Step("Open login page")
    public LoginPage openPage() {
        openPath("/login");
        return this;
    }

    @Step("Click 'Register' link under the login form")
    public RegisterPage clickRegisterLink() {
        driver.findElement(REGISTER_LINK).click();
        return new RegisterPage(driver, config);
    }

    @Step("Fill and submit form")
    public HomePage fillAndSubmitForm(String username, String password) {
        typeUsername(username);
        typePassword(password);
        return submit();
    }

    @Step("Type username: {username}")
    public LoginPage typeUsername(String username) {
        fill(LOGIN_INPUT, username);
        return this;
    }

    @Step("Type password")
    public LoginPage typePassword(String password) {
        fill(PASSWORD_INPUT, password);
        return this;
    }

    @Step("Submit login form")
    public HomePage submit() {
        driver.findElement(SUBMIT).click();
        waitUrlIsHome();
        return new HomePage(driver, config);
    }

    @Step("Submit login form expecting validation error")
    public LoginPage submitExpectingError() {
        driver.findElement(SUBMIT).click();
        waitVisible(ERROR);
        return this;
    }

    @Step("Verify embedded header is mounted")
    public LoginPage shouldShowEmbeddedHeader() {
        waitVisible(HEADER);
        return this;
    }

    @Step("Verify login form is mounted")
    public LoginPage shouldShowLoginForm() {
        waitVisible(FORM_TITLE);
        waitVisible(LOGIN_INPUT);
        waitVisible(PASSWORD_INPUT);
        waitVisible(SUBMIT);
        return this;
    }

    @Step("Verify form title message: {message}")
    public LoginPage shouldHaveFormTitle(String message) {
        waitTextContains(FORM_TITLE, message);
        return this;
    }

    @Step("Verify error message: {message}")
    public LoginPage shouldHaveErrorMessage(String message) {
        waitTextContains(ERROR, message);
        return this;
    }
}
